### PR TITLE
fix(ivy): OnDestroy hook should not be called twice for a directive

### PR DESCRIPTION
**Pull request by marclaval**
@vicb this is the issue I was fixing in the middle of the pipes PR.
I've added more tests than needed to ensure that directives' hooks are well tested.
Only the last one was failing without the fix.

### CODE REVIEW

File: packages/core/src/render3/node_manipulation.ts [hunk 1]
```diff
@@ -233,7 +233,9 @@ export function destroyViewTree(rootView: LView): void {
     }
 
     if (next == null) {
-      while (viewOrContainer && !viewOrContainer !.next) {
+      // if the viewOrContainer is the rootView, then the cleanup is done twice (e.g. ngOnChanges on
```
Comment (vicb) [inline packages/core/src/render3/node_manipulation.ts]: I don't understand the comment. Could you expand a little bit ?

**Review by kara — APPROVED** [review]
LGTM.

Note: I think the commit message is a bit confusing and should probably be changed. It would be more accurate to note that it fixes destroy hooks for directives on elements specifically. I spent a few minutes trying to figure out how your destroy hook fix could possibly affect init hooks, etc, but it looks like you were just adding missing tests.

**Comment by marclaval** [comment]
I did the requested cleanup. Sorry for the confusion, I hope it is clearer like this.

File: packages/core/test/render3/lifecycle_spec.ts [hunk 2]
```diff
@@ -423,6 +448,43 @@ describe('lifecycles', () => {
...
 
+    it('should be called on directives after component', () => {
+      /** <comp directive></comp> */
+      function Template(ctx: any, cm: boolean) {
+        if (cm) {
+          elementStart(0, Comp, null, [Directive]);
+          elementEnd();
+        }
+        Comp.ngComponentDef.h(1, 0);
+        componentRefresh(1, 0);
```
Comment (kara) [inline packages/core/test/render3/lifecycle_spec.ts]: Add directive's hostBindings and refresh calls?

Comment (marclaval) [reply packages/core/test/render3/lifecycle_spec.ts]: Added

**Review by kara — APPROVED** [review]
LGTM

**Comment by kara** [comment]
[presubmit]( http://test/OCL:186638556:BASE:186638641:1519324666169:9ff9dfdb)

**Review by vicb — APPROVED** [review]
Thanks for the update

**Pull request closed without merging**